jetty-webapp/src/test/java/org/eclipse/jetty/webapp/ForcedServletTest.java

```diff
@@ -0,0 +1,239 @@
+    {
+        StringBuilder rawRequest = new StringBuilder();
+        rawRequest.append("GET /hello.jsp HTTP/1.1\r\n");
+        rawRequest.append("Host: local\r\n");
+        rawRequest.append("Connection: close\r\n");
+        rawRequest.append("\r\n");
+
+        String rawResponse = connector.getResponse(rawRequest.toString());
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+        // status code 555 is from RejectUncompiledJspServlet
+        assertEquals(555, response.getStatus());
+    }
+
+    /**
+     * Test access of a jsp resource that does not exist in the base resource, but in the descriptor, as a precompiled jsp entry
+     */
+    @Test
+    public void testPrecompiledEntry() throws Exception
+    {
+        StringBuilder rawRequest = new StringBuilder();
+        rawRequest.append("GET /precompiled/world.jsp HTTP/1.1\r\n");
+        rawRequest.append("Host: local\r\n");
+        rawRequest.append("Connection: close\r\n");
+        rawRequest.append("\r\n");
+
+        String rawResponse = connector.getResponse(rawRequest.toString());
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+        assertEquals(200, response.getStatus());
+
+        String responseBody = response.getContent();
+        assertThat(responseBody, containsString("This is the FakePrecompiledJSP"));
+    }
+
+    public static class ForcedWebAppContext extends WebAppContext
+    {
+        @Override
+        protected void startWebapp() throws Exception
+        {
+            System.err.printf("### Thread.cl = %s%n", Thread.currentThread().getContextClassLoader());
+
+            // This will result in a 404 for all requests that don't belong to a more precise servlet
+            forceServlet("default", ServletHandler.Default404Servlet.class);
+            addServletMapping("default", "/");
+
+            // This will result in any attempt to use an JSP that isn't precompiled and in the descriptor with status code 555
+            forceServlet("jsp", RejectUncompiledJspServlet.class);
+            rejectForcedJspPaths();
+            addServletMapping("jsp", "*.jsp");
+
+            super.startWebapp();
+        }
+
+        private void addServletMapping(String name, String pathSpec)
+        {
+            ServletMapping mapping = new ServletMapping();
+            mapping.setServletName(name);
+            mapping.setPathSpec(pathSpec);
+            getServletHandler().addServletMapping(mapping);
+        }
+
+        private void rejectForcedJspPaths()
+        {
+            for (ServletHolder h : getServletHandler().getServlets())
+            {
+                if (h.getForcedPath() != null)
+                {
+                    System.err.println("FORCED " + h.getForcedPath());
+                    h.setHeldClass(RejectUncompiledJspServlet.class);
+                    h.setForcedPath(null);
+                }
+            }
+        }
```

Comment:
Remove this method as that is the work around to go into the clients usage rather than the release.
```suggestion
```